lib/dry/types/hash/key.rb

```diff
@@ -0,0 +1,60 @@
+require 'dry/equalizer'
+
+module Dry
+  module Types
+    class Hash < Definition
+      class Key
+        include Type
+        include ::Dry::Equalizer(:name, :type, :options)
+        include Decorator
+        include Builder
+
+        attr_reader :name
+
+        def initialize(type, name, required: Undefined, **options)
+          required = Undefined.default(required) do
+            type.meta.fetch(:required) { !type.meta.fetch(:omittable, false) }
+          end
```

Comment:
I see down at the bottom of the diff on this PR (in `#transform_key_type`) that you have:

```ruby
key.required(false) if omittable?(constructor)
```

Is your thinking (either as later work in this PR, or in another subsequent one) that you might provide another way to provide _individually_ omittable keys _without_ having to rely on a meta attribute?  IIRC that was part of the bug we discovered in rom where it was exporting read hash schema ASTs with the metas excluded.

Reply:
this is part of the plan, but I don't have all pieces in mind yet. It seems reasonable to support at least this way of defining omittable keys
```ruby
Types::Hash.schema(
  required: Types::Bool,
  optional?: Types::Bool
)
```

Another possible API is building a key object out of any type:

```ruby
Types::Strict::String.key(:name, required: false) # => #<Types::Hash::Key>
```

This is flexible, but because of OOP it tangles things a bit. 
Now still, we're yet to figure out if anything of this can help with that ROM issue. For that we first need to extract its description from the gitter archives :)

Reply:
@flash-gordon you're both and persuasive, I'll try do that this weekend :P

Reply:
Apart from that, good to hear that it's part of your plan to support a better first-class ways of declaring omittable keys.